### ISSUE TITLE
Effective ACL Resource argument requires 2n+1 requests

### ISSUE THREAD

**Issue by bblfish** [solid/web-access-control-spec#99]
The Access Control Panel has been putting together protocol comparisons between WAC, ACP, and various WAC+ solutions. (WAC+ are potential ways of extending WAC in a backward compatible way).

The [Effective Access Control Resource Discovery Use Case](https://github.com/solid/authorization-panel/blob/main/proposals/evaluation/uc-0-effective-acr.md) reveals a quite dramatic problem for WAC when a resource is using a default rule further up the hierarchy. We have 4 WAC+ ways to solve the problem listed in the document there. Please could the WAC community help us select one of them?

## Problem description

Our resources are set up as follows:

```turtle  
</default.acr>
</foo/bar/baz/x>
</foo/bar/baz/x.acr>
```
0. The client makes a request on `</foo/bar/baz/x>` and receives a response  —

```HTTP
401 Credentials Required
Link: </foo/bar/baz/x.acr>; rel="acl"
Link: <.>; rev="http://www.w3.org/ns/ldp#contains"
...
```

The server needs to link to the closest possible ACR for the resource, to allow a client with Control rights to know what  ACR to edit. There is otherwise no way for the client to know where the ACR should be placed.

1. The client makes a request to the ACR `</foo/bar/baz/x.acr>` but receives the reponse —

```HTTP
404 Not Found
```
— as the resource does not yet exist.

2. First the client is lucky enough to be shown the reverse `ldp:contains` relation in the response in (1) above, so it can do a HEAD on that to find its `ACL`.
```HTTP
HEAD /foo/bar/baz/ HTTP/1.1
```
— and with luck the server will respond —
```HTTP
200 Ok
Link: </foo/bar/baz/.acr>; rel="acl"
Link: </foo/bar/>; rev="http://www.w3.org/ns/ldp#contains"
```

3.  The client can then continue with —
```HTTP
GET /foo/bar/baz/.acr HTTP/1.1
```
— to which the server will also return —
```HTTP
404 Not Found
```
— as the resource does not yet exist.

4. As a result, the client will need to look up one level in the hierarchy to search for the effective ACR:
```HTTP
HEAD /foo/bar/ HTTP/1.1
```
— and with luck the server will respond —
```HTTP
200 Ok
Link: </foo/bar/.acr>; rel="acl"
Link: </foo/>; rev="http://www.w3.org/ns/ldp#contains"
```

5. The client can then continue with —
```HTTP
GET /foo/bar/.acr HTTP/1.1
```
— to which the server will also return —
```HTTP
404 Not Found
```
— as the resource does not yet exist.

6. As a result, the client will need to look up one level in the hierarchy to search for the effective ACR:
```HTTP
HEAD /foo/ HTTP/1.1
```
— and with luck the server will respond —
```HTTP
200 Ok
Link: </foo/.acr>; rel="acl"
Link: </>; rev="http://www.w3.org/ns/ldp#contains"
```

7. The client can then continue with —
```HTTP
GET /foo/.acr HTTP/1.1
```
— to which the server will also return —
```HTTP
404 Not Found
```
— as the resource does not yet exist.

8. As a result, the client will need to look up one level in the hierarchy to search for the effective ACR:
```HTTP
HEAD / HTTP/1.1
```
— and with luck the server will respond —

```HTTP
200 Ok
Link: </.acr>; rel="acl"
```

9. The client can then continue with —
```HTTP
GET /.acr HTTP/1.1
```
— to which the server will finally return the content.

The number of requests needed to find the default can be calculated by the formula —
```
reqN = 2 * slashes + extraFile
```
— where —
* `slashes` is the number of slashes between the original resource and the effective ACR
* `extraFile` is `0` if the resource is an `ldp:Container` and `1` if it is a plain resource

So in our example above we have `reqN = 2 * 4 + 1 = 9`.

What we want is for that number to be cut down to 1 request. 
As I mentioned there are 3 or 4 answers. 
The one I prefer, as it is the simplest is explained in [define another relation for non-created ACR](https://github.com/solid/authorization-panel/issues/248).

**Comment by kjetilk** [comment]
I believe and always have, that the resource should link directly to its effective access control resource. It may also link to a resource that could be created to control it, but that is of secondary value, I think.

**Comment by bblfish** [comment]
> I believe and always have, that the resource should link directly to its effective access control resource. It may also link to a resource that could be created to control it, but that is of secondary value, I think.

@kjetilk I agree for clients only interested in reading the ACL, which should be 99.99% of the cases if not more, the only interesting acl is the effective access control resource. So I think we can assume that that is what "acl" links refer to.

**But** - and this is really the crux of the matter - if you only link to that effective resource, say the default root ACL `</default.acr>`, then the client ACL editor cannot change the rules if it wants to make new defaults higher up the directory tree: because it will not know what the ACL of that resource is!

So that is why I think the really simple answer is: we should link to both, but distinguish not-yet existing resources from the active ones. I don't know what to call that link. "access-control", "potential-acr", "specific acr", "my-rule-resources"... I don't know....

The advantage of having two links is that 
1) people who want their servers with most of the rules in a few default places, get that working with minimal number of such resources. This could be very useful for servers where most of everything is either public or private.
2) people who prefer to have very specific rules tightly specified for one single resource can have that too. Our use case [from the use case document](https://solid.github.io/authorization-panel/authorization-ucr/#inheritance-readcreate) is actually one that starts off with a generic default and then moves to a more specific rule for a given container.

**Comment by kjetilk** [comment]
> **But** - and this is really the crux of the matter - if you only link to that effective resource, say the default root ACL `</default.acr>`, then the client ACL editor cannot change the rules if it wants to make new defaults further into the directory tree: because it will not know what the URI of the ACL for that resource is or should be. We need the new link type to help the client know what link to follow in which circumstance.

It might be that I don't understand what you are referring to, but I think that it can, given the slash semantics and container hierarchy assumption of Solid. Given that, it is, as far as I can tell, perfectly legitimate for the client to parse the URI of the access control resource and the target URI, and reason that they can create a new access control resource somewhere in between, if they are permitted to.

Therefore, I think that only a single link is required, a link targeting the effective access control resource, but I'd be OK with adding more, as being explicit about it is often a good practice.

**Comment by bblfish** [comment]
@kjetilk 
> but I'd be OK with adding more, as being explicit about it is often a good practice.

Great [1]. We would just need to come up with a good name for the missing relation.

Henry 

[1] I'd say that working with decentralised hyper-apps requires a very high degree of explicitness. In usual client/server apps a lot can be implicit, because the person building the client also controls the server, and the location of data on the server as well as the schemas used. Here all we have to agree on is a protocol, and some Linked Data Vocabularies. The type of data published, the way people from different cultures prefer to organise their data and the schemas they will use all has to be left open.

**Comment by kjetilk** [comment]
> [1] I'd say that working with decentralised hyper-apps requires a very high degree of explicitness.

I absolutely agree with that, in the interest of supporting HATEOAS, being explicit is very important. 

The reason why am willing to depart from that ideal in this case is that in principle, the client can put the access controls at any level between the effective access control and itself. It all depends on what the client wants, and so, having that as out-of-band knowledge seems OK to me in this case.

**Comment by bblfish** [comment]
> > **But** - and this is really the crux of the matter - if you only link to that effective resource, say the default root ACL `</default.acr>`, then the client ACL editor cannot change the rules if it wants to make new defaults further into the directory tree: because it will not know what the URI of the ACL for that resource is or should be. We need the new link type to help the client know what link to follow in which circumstance.
> 
> It might be that I don't understand what you are referring to, but I think that it can, given the slash semantics and container hierarchy assumption of Solid. Given that, it is, as far as I can tell, perfectly legitimate for the client to parse the URI of the access control resource and the target URI, and reason that they can create a new access control resource somewhere in between, if they are permitted to.

Say the a `GET` on `</foo/bar/baz/secret>` returns

```HTTP
401 Credentials Required
WWW-Authenticate: Http-Sig
Link: <.>; rev="http://www.w3.org/ns/ldp#contains"
Link: </default.acl>; rel="acl"
...
```

Indeed, this is the best for auth based clients who want to find out what rights they may have for the given resource, as the `acl` link points them right to the effective resource.  I suggest we should keep things that way. 

In that situation access to `</foo/bar/baz/secret>` will be controlled by the  rules in `</default.acl>`. If the user wants to change the default for that resource only, how is the client going to know where the `access control resource for `</foo/bar/baz/secret>` should be placed? Where should it `PUT` the new rules?

Should it just guess and try 
 * `</foo/bar/baz/secret.acr>`
 * `</foo/bar/baz/secret.acl>`
 * `</foo/bar/baz/secret.ac>`
 * `</foo/bar/baz/secret.control>`
 * `</foo/bar/baz/secret.kontrol>`
 * `</foo/bar/baz/secret.ctrl>`
 * `</foo/bar/baz/secret?ctrl>`
 * `</foo/bar/baz/secret?control=1>`
 * `</foo/bar/baz/secret?ctrl=sdj3j2dd342a>` //<= for servers that don't want anyone to guess the URL

This would be solved if the server returned 

```HTTP
401 Credentials Required
WWW-Authenticate: Http-Sig
Link: <.>; rev="http://www.w3.org/ns/ldp#contains"
Link: </default.acl>; rel="acl"
Link: </foo/bar/baz/secret?ctrl>; rel="direct-control"
...
```
 
Of course we need a word better than "direct-control".

The other simple answer where "acl" points directly to the direct access control resource, then we have the problem of clients having to find the effective one as show in the main text of this issue, leading to a problem of 2n+1 connections.

@kjetilk wrote
> The reason why am willing to depart from that ideal in this case is that in principle, the client can put the access controls at any level between the effective access control and itself. It all depends on what the client wants, and so, having that as out-of-band knowledge seems OK to me in this case.

True, and that is why, in a way, I think that there is some degree of usefulness in allowing an Access Control Editor client to follow the container hierarchy to choose the right level at which to place the rules. 

But that still requires the intermediate containers to have a link to their non-default ACL.
So say the client in our example wanted not only to change the rules for `</foo/bar/baz/secret>` as described above, but also set a new default for `</foo/bar/baz/>`, then it would want to make a `HEAD` request on `</foo/bar/baz/>` in order to find out what the URL of its acl should be. Clearly we'd have the same problem of the client having to guess the associated acl resource, if that is not made explicit.

**Comment by kjetilk** [comment]
Mmmm, right! We need a more open brainstorming session, I think.

Indeed, I think it should be up to the client to choose where in the hierarchy it is most suitable to put access control (subject to control permission of course), and so, it doesn't make sense for the server to enumerate all possibilities in every header for every request. 

I have also been thinking that we may be using headers a little too much and the body possibly not enough. I'm thinking that the container may contain more descriptive resources in RDF. If we need to enumerate affordances, it might be better done in the body. If the container is unsuited for it, perhaps the payload of an `OPTIONS` request on the target resource is better? Or perhaps we could define some patterns that can be used for the client to put a valid access control into the hierarchy suitably.

I very much agree we need to avoid _2n+1_ number of requests for this kind of behavior.

**Comment by bblfish** [comment]
> Indeed, I think it should be up to the client to choose where in the hierarchy it is most suitable to put access control (subject to control permission of course), and so, it doesn't make sense for the server to enumerate all possibilities in every header for every request.

I also agree with that now - I had not really thought of that before looking at `:default` acls to be honest.

If we adopt the two link relation types then we would have the following link headers 
for each of the relevant resources:

|  resource path      | resource's ACR (exists?)|  Active  ACR "acl"                    |
|------------------|---------------------|-----------------------|
| `/foo/bar/baz/x`   |  `/foo/bar/baz/x.ac`   (❌)  | `/default.ac`                  |
| `/foo/bar/baz/`     | `/foo/bar/baz/.ac`     (❌)  | `/default.ac`                  |
| `/foo/bar/`             | `/foo/bar/.ac`     (❌)  | `/default.ac`                  |
| `/foo/`     | `/foo/.ac`   (❌)    | `/default.ac`                  |
| `/`            | `/default.ac`     (✅)  | `/default.ac`                  |

The idea here is that every resource has a Link to it's "Access Control Resource" ACR to use the terminology from ACP. But as opposed to ACP that resource would not have to exist. We see above that only the Access Control Resource that exists is the `</default.ac>` resource. All others resources in the second column would return a `405 Method Not Allowed` on a GET request, and specify that only PUT and PATCH are allowed.

So now on making a request to `/foo/bar/baz/x` an access control editor would know that it could edit `/foo/bar/baz/x.ac` to affect that resource along. But it would know that it could also edit the ACRs for the intermediate containers if it wanted the rules it sets to have a wider range of uses. How would it find those intermediate containers? Well the answer is simple: just the way I showed in the issue description, by following `Link: <..>; rev="http://www.w3.org/ns/ldp#contains"`links. That may require going through the directory hierarchy to **SET** (`PUT` or `PATCH`) a new ACL. But that is fine as setting an acl should be a very rare phenomenon compared to reading the ACL. 

So let us imagine that the Client wants to set an acl default at `/foo/bar/`. After doing a HEAD on `/foo/bar/baz/` it would find that `/foo/bar/` was the grandparent container, and a HEAD on that would tell it that `/foo/bar/.ac` was the ACR for that container (but not the active one). So it could do a `PUT` on `/foo/bar/.ac` with the new rules and the situation would be as follows:

|  resource path      | resource's ACR (exists?)|  Active  ACR "acl"                    |
|------------------|---------------------|-----------------------|
| `/foo/bar/baz/x`   |  `/foo/bar/baz/x.ac`   (❌)  | `/foo/bar/.ac`                  |
| `/foo/bar/baz/`     | `/foo/bar/baz/.ac`     (❌)  | `/foo/bar/.ac`                  |
| `/foo/bar/`             | `/foo/bar/.ac`     (✅)  | `/foo/bar/.ac`                  |
| `/foo/`     | `/foo/.ac`   (❌)    | `/default.ac`                  |
| `/`            | `/default.ac`     (✅)  | `/default.ac`                  |

Note that whenever a resource's ACR has a ✅ then the resource need only present the "acl" link relation.

**Comment by csarven** [comment]
(mentioned this before in meetings, issues..)

First: There is an issue on this https://github.com/solid/specification/issues/106 but I see no reference to it in proposed discussions. As there was no consensus on 106, I simplified the effective-acl-resource-algorithm so that it can be used by both servers and clients. And, then added a section in WAC ED to acknowledge *Distinct Effective ACL Resource Extensions*: https://solid.github.io/web-access-control-spec/#distinct-effective-acl-resource-extensions -- no reference to that either. So, the good news is that, it is possible. That section wasn't a random thing thrown in ;)

Second: For the time being, I'm taking the 2n+1 formula (but discussed in terms of complexity) with a grain of salt. WAC doesn't prohibit a server to have a representation for every ACL resource. It only says that it should not have a representation in order for defaults to be used from a container higher up (towards root). So, yes, a particular server implementation with a particular data organisation will encounter 2n+1 as worst case but none of that is factored in the discussion. It is based off an example - I'm not arguing that it is not a legitimate case but nevertheless, it is a case. Can we have proper best, average, worst cases based on data/time/space, if we are going to talk about complexity?

**Comment by kjetilk** [comment]
I don't think anybody is suggesting that you didn't capture the consensus at the time, @csarven , but it is worth taking a look back at certain aspects. 

FWIW, in this very open world, it is quite likely that someone will stumble on the worst case and the practical average case is basically _a priori_ unknowable. I therefore think it is quite sufficient to consider the worst case unless there is obviously no use case for the situation where it arises. That is not to say that it is the only consideration, but it should be an important one.

**Comment by bblfish** [comment]
## Prior work

Thanks @csarven for pointing to [spec #106](https://github.com/solid/specification/issues/106) and prior work. As I see we have quite a number of people who are for a header to point to the effective acl: @dmitrizagidulin, @RubenVerborgh, @kjetilk, @acoburn (suggested using `http://www.trellisldp.org/ns/trellis#effectiveAcl`) so there is strong consensus on having two URLs one for the resources own ACR and one for the effective one. Indeed the new spec, as you point out, states:

> To enable clients to perform discovery faster, separate link relation type targeting the effective ACL resource is allowed, but no behaviour is defined by this specification

So what we are asking for here, is for that consensus to be realised: I.e for a relation name to be coined so that all clients can follow that link, so as to avoid them having to make 2n+1 requests in the case where defaults are specified.
Whether the current "acl" header should point to the effective ACR and a new relation to the existing ACL, or vice versa, is I think something that can be established by looking at how implementations are working.

##  On 2n+1

2n+1 is precisely the number of requests a client will have to make if it is to follow the procedure specified when dealing with default rules. That is a consequence of having default rules that are overridable by creating a new ACR, and I think  now that there are reasonable use cases for that, such as a public web server where all content is publicly readable: why bother having ACRs on each resource, when there is only one effective one. I would complement that with [:imports](https://github.com/solid/authorization-panel/issues/210) to allow new acls to avoid having to copy new rules everywhere. But that is a different issue.

**Comment by bblfish** [comment]
I don't think we are at the point of having to decide on MUST or MAY. 
We are at the point of wanting a name for the relation, either a URL or a short name to go with "acl".

I think having the new relation does not invalidate the "effective-act-resource algorithm", since a Control-Client may want to edit intermediate ACLs.

**Comment by acoburn** [comment]
Whether there exists a link to an "effective ACL" or not, when inheritance is in effect with WAC, any client needs to understand _a lot_ about the overall (global) structure of the ACLs involved in a hierarchy of resources in order to make decisions about how a change in that effective ACL would affect the permission structure of that hierarchy. And while in the context of Solid there is a Pod owner which does have (or can have) global access to the ACL structure, that is not an assumption that holds generally.

And if a client does not have global understanding of the ACL structure for a hierarchically arranged set of resources, it doesn't matter whether that client can efficiently navigate to an "effective ACL" or not: that client _still_ needs to navigate through the entire tree (which may not even be a possibility) in order to understand what will happen if that "effective ACL" is changed.

In a word: this just makes "unintended consequences" much, much more likely for clients. I would tread _very_ carefully here.

**Comment by acoburn** [comment]
> Well it will save 2n+1 requests to find them, that I made crystal clear above. There is no doubt about it.

Indeed. And now you need to traverse the entire tree downwards to find out what the access controls in that resource actually mean. And those resources you traverse downward have already been fetched when you traverse _upwards_.

**Comment by acoburn** [comment]
Consider the structure:

```
/
  container/ *
     child1/  *
        grandchild1a
        grandchild1b
        grandchild1c *
    child2/
        grandchild2a
        grandchild2b *
        grandchild2c
    child3/
        grandchild3a
        grandchild3b
        grandchild3c *
```

There is an acl on all resources with an asterisk. Say you perform a GET on `/container/child3/grandchild3b`

You get back two link headers: one pointing to the (non-existent) ACL for `/container/child3/grandchild3b` and one pointing to the ACL on `/container/`. What does the client do if it wants to give a user read access to `/container/child3/grandchild3b`?

One option:
* Create a new ACL resource associated with `/container/child3/grandchild3b` and add that user there. This is simple and can be done without the new link header.

The alternative that this new link header enables:
* Adjust the ACL associated with `/container/`. 

However, what is the side-effect of making that second change? This user is now granted access to all of the other resources under `/container/` that don't already have an ACL. How does the client know that? By traversing the entire tree.
